README.md
<h1 align="center">
  <img src="https://avatars.githubusercontent.com/u/54333248?s=200&v=4">
    <br>
    Pinecone Client V3
    <br>
</h1>

A python client for Pinecone. For more information, see the docs at https://www.pinecone.io/docs/

Pinecone Client V3 is based on pre-compiled Rust code and gRPC, aimed at improving performance and stability.
Using native gRPC transport, client V3 is able to achieve a 2x-3x speedup for vector upsert over the previous Rest-based client versions, as well as a 10-20% speedup for vector query latency.
As the client installation is fully self-contained, it does not require any additional dependencies (e.g. `grpcio`), making it easier to install and use in any python environment.

> **_⚠️ Warning_**
>
> This is a **public preview** ("Beta") version. Please test it thoroughly before using it in production.    
> Some functionalities are not backwards compatible with the previous versions of the client, please refer to the [cahangelog]() for more details.

## Installation

Install public preview version from pip:
```shell
pip3 install pinecone-client==3.0.0rc2
```

## Building from source and contributing
See [CONTRIBUTING.md](CONTRIBUTING.md)

## Migrating from pinecone client V2
If you are migrating from pinecone client V2, here is the most minimal code change required to upgrade to V3:
```python
# Pinecone client V2
import pinecone
pinecone.init() # One time init
...
index = pinecone.Index("example-index")
index.upsert(...)

# Pinecone client V3
from pinecone import Client
pinecone = Client() # This is now a `Client` instance
...
# Unchanged!
index = pinecone.Index("example-index")
index.upsert(...)
```
For more API changes see [CHANGELOG.md](CHANGELOG.md)

# Usage
## Control plane operations
### Creating a Client instance
The `Client` is the main entry point for control operations like creating, deleting and configuring pinecone indexes.  
Initializing a `Client` requires your Pinecone API key and a region, which can be passed as either environment variables or as parameters to the `Client` constructor.

```python
import os
from pinecone import Client

# Initialize a client using environment variables
os.environ['PINECONE_API_KEY'] = 'YOUR_API_KEY'
os.environ['PINECONE_REGION'] = 'us-west1-gcp'
client = Client()

# Initialize a client using parameters
client = Client(api_key = 'YOUR_API_KEY', region = 'us-west1-gcp')
```

### Creating an index

The following example creates an index without a metadata configuration.  

By default, all metadata fields are indexed.

```python

from pinecone import Client

client = Client(api_key="YOUR_API_KEY", region="us-west1-gcp")

index = client.create_index("example-index", dimension=1024)
```

If some of the metadata fields contain data payload such as raw text, indexing them would make the Pinecone less efficient. In such cases, it is recommended to configure the index to only index specific metadata fields which are used for query filtering.  

The following example creates an index that only indexes the `"color"` metadata field. 

```python
metadata_config = {
    "indexed": ["color"]
}

client.create_index("example-index-2", dimension=1024,
                      metadata_config=metadata_config)
```

#### Listing all indexes

The following example returns all indexes in your project.

```python
active_indexes = client.list_indexes()
```

#### Getting index configuration

The following example returns information about the index `example-index`.

```python
index_description = client.describe_index("example-index")
```

#### Deleting an index

The following example deletes `example-index`.

```python
client.delete_index("example-index")
```

#### Scaling an existing index number of replicas

The following example changes the number of replicas for `example-index`.

```python
new_number_of_replicas = 4
client.scale_index("example-index", replicas=new_number_of_replicas)
```
## Data plane operations
### Creating an Index instance
The index object is the entry point for data operations like upserting, querying and deleting vectors.
```python
from pinecone import Client
client = Client(api_key="YOUR_API_KEY", region="us-west1-gcp")
index = client.get_index("example-index")

# Backwards compatibility
index = client.Index("example-index")
```

#### Printing index statistics

The following example returns statistics about the index `example-index`.

```python
from pinecone import Client

client = Client(api_key="YOUR_API_KEY", region="us-west1-gcp")
index = client.Index("example-index")

print(index.describe_index_stats())
```


#### Upserting vectors

The following example upserts vectors to `example-index`.

```python
from pinecone import Client, Vector, SparseValues
client = Client(api_key="YOUR_API_KEY", region="us-west1-gcp")
index = client.get_index("example-index")

upsert_response = index.upsert(
    vectors=[
        ("vec1", [0.1, 0.2, 0.3, 0.4], {"genre": "drama"}),
        ("vec2", [0.2, 0.3, 0.4, 0.5], {"genre": "action"}),
    ],
    namespace="example-namespace"
)

# Mixing different vector representations is allowed
upsert_response = index.upsert(
    vectors=[
        # Tuples 
        ("vec1", [0.1, 0.2, 0.3, 0.4]),
        ("vec2", [0.2, 0.3, 0.4, 0.5], {"genre": "action"}),
        # Vector objects
        Vector(id='id1', values=[1.0, 2.0, 3.0], metadata={'key': 'value'}),
        Vector(id='id3', values=[1.0, 2.0, 3.0], sparse_values=SparseValues(indices=[1, 2], values=[0.2, 0.4])),
        # Dictionaries
        {'id': 'id1', 'values': [1.0, 2.0, 3.0], 'metadata': {'key': 'value'}},
        {'id': 'id2', 'values': [1.0, 2.0, 3.0], 'sparse_values': {'indices': [1, 2], 'values': [0.2, 0.4]}},
    ],
    namespace="example-namespace"
)
```

#### Quering an index by a new unseen vector

The following example queries the index `example-index` with metadata
filtering.

```python
query_response = index.query(
    values=[1.0, 5.3, 8.9, 0.5], # values of a query vector
    sparse_values = None, # optional sparse values of the query vector
    top_k=10,
    namespace="example-namespace",
    include_values=True,
    include_metadata=True,
    filter={
        "genre": {"$in": ["comedy", "documentary", "drama"]}
    }
)
```

#### Quering an index by an existing vector ID

The following example queries the index `example-index` for the `top_k=10` nearest neighbors of the vector with ID `vec1`.

```python
query_response = index.query(
    id="vec1",
    top_k=10,
    namespace="example-namespace",
    include_values=True,
    include_metadata=True,
)
```

#### Deleting vectors

```python
# Delete vectors by IDs 
index.delete(ids=["vec1", "vec2"], namespace="example-namespace")

# Delete vectors by metadata filters
index.delete_by_metadata(filter={"genre": {"$in": ["comedy", "documentary", "drama"]}}, namespace="example-namespace")

# Delete all vectors in a given namespace (use namespace="" to delete all vectors in the DEFAULT namespace)
index.delete_all(namespace="example-namespace")
```

#### Fetching vectors by ids

The following example fetches vectors by ID without querying for nearest neighbors.

```python
fetch_response = index.fetch(ids=["vec1", "vec2"], namespace="example-namespace")
```


#### Update vectors

The following example updates vectors by ID.

```python
update_response = index.update(
    id="vec1",
    values=[0.1, 0.2, 0.3, 0.4],
    set_metadata={"genre": "drama"},
    namespace="example-namespace"
)
```
# Performance tuning for upserting large datasets
To upsert an entire dataset of vectors, we recommend using concurrent batched upsert requests. The following example shows how to do this using the `asyncio` library:
```python
import asyncio
from pinecone import Client, Vector

def chunker(seq, batch_size):
    return (seq[pos:pos + batch_size] for pos in range(0, len(seq), batch_size))

async def async_upload(index, vectors, batch_size, max_concurrent=50):
    sem = asyncio.Semaphore(max_concurrent)
    async def send_batch(batch):
        async with sem:
            return await index.upsert(vectors=batch, async_req=True)
    
    await asyncio.gather(*[send_batch(chunk) for chunk in chunker(vectors, batch_size=batch_size)]) 

# To use it:
client = Client()
index = client.get_index("example-index")
asyncio.run(async_upload(index, vectors, batch_size=100))

# In a jypter notebook, asyncio.run() is not supported. Instead, use
await async_upload(index, vectors, batch_size=100)  
```

#### TODO: Decide if we want to suggest this longer, more verbose version, which includes a progress bar and return type: 

```python
from tqdm.asyncio import tqdm
import asyncio
from pinecone import UpsertResponse, Client, Vector

def chunker(seq, batch_size):
    return (seq[pos:pos + batch_size] for pos in range(0, len(seq), batch_size))

async def async_upload(index, vectors, batch_size, max_concurrent=50):
    sem = asyncio.Semaphore(max_concurrent)
    async def send_batch(batch):
        async with sem:
            return await index.upsert(vectors=batch, async_req=True)

    tasks = [send_batch(chunk) for chunk in chunker(vectors, batch_size=batch_size)]
    pbar = tqdm(total=len(vectors), desc="upserted vectors")
    total_upserted_count = 0
    for task in asyncio.as_completed(tasks):
        res = await task
        total_upserted_count += res.upserted_count
        pbar.update(res.upserted_count)
    return UpsertResponse(upserted_count=total_upserted_count)

# To use it:
client = Client()
index = client.get_index("example-index")
vectors = [Vector(id=f"vec{i}", values=[1.0, 2.0, 3.0]) for i in range(1000)]
res = asyncio.run(async_upload(index, vectors, batch_size=100))

# In a jypter notebook, asyncio.run() is not supported. Instead, use
res = await async_upload(index, vectors, batch_size=100)  
```

# Limitations

## Code completion and type hints
Due to limitations with the underlying `pyo3` library, code completion and type hints are not available in some IDEs, or might require additional configuration.
- **Jupyter notebooks**: Should work out of the box
- **VSCode**: Change the [`languageServer`](https://code.visualstudio.com/docs/python/settings-reference#_intellisense-engine-settings) setting to `jedi`
- **PyCharm**: For the moment, all fucntion signatures would show `(*args, **kwargs)`. We are working on a solution ASAP. (Function docstrings would still show full arguments and type hints)


CHANGELOG.md
# Changelog

## 3.0.0rc1 - breaking changes!!
**New client architecture based on pre-compiled Rust code and gRPC**
### Breaking changes and migration guide:
#### Client initialization
- The `Client` object is the main entry point for control operations like creating, deleting and configuring pinecone indexes. 
    It replaces the global `pinecone.init()` function from previous versions.    
    For most minimal code change, you can initialize a `Client` instance and use it as a drop-in replacement for the global `pinecone` object (see [README](https://github.com/pinecone-io/pinecone-client#migrating-from-pinecone-client-v2)).
- The `envrionement` parameter for `Client` initialization was renamed `region` instead. Similarly, the `PINECONE_ENVIRONMENT` environment variable was renamed to `PINECONE_REGION` as well.

#### Control plane operations
- The `create_index()` method is mostly unchanged, but a few deprecated or unused parameters were removed: `index_type`, `index_config`
- The `configure_index()` method was removed. Use `scale_index()` instead.

#### Upsert operation
- `index.upsert()` now supports mixing vector represenations in the same batch (see [Upserting vectors](https://github.com/pinecone-io/pinecone-client#upserting-vectors)
- When used with `async_req=True`, `index.upsert()` now returns an `asyncio` coroutine instead of a `concurrent.futures.Future` object. See [Performance tuning](https://github.com/pinecone-io/pinecone-client#performance-tuning-for-upsering-large-datasets) for more details.

#### Query operation
- Querying using an existing vector id was separated into a new method `index.query_by_id()`. The `index.query()` method now only accepts a vector values (and optioanl sparse values).
- The both `index.query()` and `index.query_by_id()` now return a list of `QueryResult` objects. The `QueryResult` object has the following attributes:
    - `id` - the vector id
    - `score` - the ANN score for the given query result
    - `values` - optional vector values if `include_values=True` was passed to the query method
    - `sparse_values` - optional vector sparse values if `include_values=True` was passed to the query method
    - `metadata` - optional vector metadata if `include_metadata=True` was passed to the query method
    - `to_dict()` - a method that returns a dictionary representation of the query result



pinecone/src/index.rs
use crate::data_types::convert_upsert_enum_to_vectors;
use crate::data_types::UpsertRecord;
use crate::utils::errors::{PineconeClientError, PineconeResult};
use client_sdk::data_types as core_data_types;
use client_sdk::index as core_index;
use client_sdk::utils::errors::PineconeClientError as core_error;
use pyo3::prelude::*;
use std::collections::{BTreeMap, HashMap};
use tokio::runtime::Handle;

#[pyclass]
pub struct Index {
    inner: core_index::Index,
    runtime: Handle,
}

impl Index {
    pub fn new(inner: core_index::Index, runtime: Handle) -> Self {
        Self { inner, runtime }
    }
}

#[pymethods]
impl Index {
    pub fn __repr__(&self) -> String {
        format!("Index: \"{name}\"", name = self.inner.name)
    }

    #[pyo3(signature = (vectors, namespace="", async_req=false))]
    #[pyo3(text_signature = "(vectors, namespace='', async_req=False)")]
    /// The `Upsert` operation writes vectors into a namespace.
    /// If a new value is upserted for an existing vector id, it will overwrite the previous value.
    ///
    /// Args:
    ///     vectors (Union[List[Tuple[str, List[float]]], List[Dict[str, Any]], List[Vector]]): A list of vectors to upsert.
    ///         A vector can be represented by:
    ///         - A `Vector` object.
    ///         - A tuple of the form (id: str, vector: List[float]) or (id: str, vector: List[float], metadata: Dict[str, Union[str, float, int, bool, List[str]]]])
    ///         - A dictionary with the keys 'id' (str), 'values' (List[float]), 'sparse_values' (optional dict in the format {'indices': List[int], 'values': List[float]}), 'metadata' (Optional[Dict[str, Any]])
    ///         Note: sparse values are not supported when using a tuple. Please use a dictionary or a `Vector` object instead.
    ///
    ///     namespace (Optional[str]): Optional namespace to which data will be upserted.
    ///     async_req (bool): When set to True, the upsert request will be performed asynchronously, and a "future" (asyncio coroutine) will be returned.
    ///
    /// Examples:
    ///     ```python
    ///     index.upsert([ Vector(id='id1', values=[1.0, 2.0, 3.0], metadata={'key': 'value'}),
    ///                    Vector(id='id3', values=[1.0, 2.0, 3.0], sparse_values=SparseValues(indices=[1, 2], values=[0.2, 0.4])) ])
    ///
    ///     index.upsert([ ('id1', [1.0, 2.0, 3.0], {'key': 'value'}),
    ///                    ('id2', [1.0, 2.0, 3.0]) ])
    ///
    ///     index.upsert([ {'id': 'id1', 'values': [1.0, 2.0, 3.0], 'metadata': {'key': 'value'}},
    ///                        {'id': 'id2', 'values': [1.0, 2.0, 3.0], 'sparse_values': {'indices': [1, 2], 'values': [0.2, 0.4]}} ])
    ///
    ///     # Mixing different vector representations is also allowed
    ///     index.upsert([ {'id': 'id1', 'values': [1.0, 2.0, 3.0], 'metadata': {'key': 'value'}, 'sparse_values': {'indices': [1, 2], 'values': [0.2, 0.4]}},
    ///                    ('id2', [1.0, 2.0, 3.0]), ])
    ///     ```
    ///
    /// Returns:
    ///     - If `async_req=False`:
    ///         UpsertResponse: An upsert response object. Currently has an 'upserted_count' field with vector count. Might be extended in the future.
    ///     - If `async_req=True`:
    ///         An `asyncio` coroutine that can be awaited using `await` or `asyncio.gather()`.
    pub fn upsert<'a>(
        &mut self,
        py: Python<'a>,
        vectors: Vec<UpsertRecord>,
        namespace: &'a str,
        async_req: bool,
    ) -> PyResult<&'a PyAny> {
        // According to tonic's documentation, cloning the generated client is actually quite cheap,
        // and that's the recommended behavior: https://docs.rs/tonic/latest/tonic/transport/struct.Channel.html#multiplexing-requests
        let mut inner_index = self.inner.clone();

        let namespace = namespace.to_owned();
        let vectors_to_upsert =
            convert_upsert_enum_to_vectors(vectors).map_err(PineconeClientError::from)?;

        if async_req {
            pyo3_asyncio::tokio::future_into_py(py, async move {
                let res = inner_index
                    .upsert(&namespace, &vectors_to_upsert, None)
                    .await
                    .map_err(PineconeClientError::from)?;
                Ok(res)
            })
        } else {
            pyo3_asyncio::tokio::get_runtime().block_on(async move {
                let res = inner_index
                    .upsert(&namespace, &vectors_to_upsert, None)
                    .await
                    .map_err(PineconeClientError::from)?;
                Ok(res.into_py(py).into_ref(py))
            })
        }
    }

    #[pyo3(signature = (top_k, values=None, sparse_values=None, namespace="", filter=None, include_values=false, include_metadata=false))]
    #[pyo3(
        text_signature = "($self, top_k, values=None, sparse_values=None, namespace='', filter=None, include_values=False, include_metadata=False)"
    )]
    /// Query
    ///
    /// The `Query` operation searches a namespace, using a query vector.
    /// It retrieves the ids of the most similar items in a namespace, along with their similarity scores.
    /// To query by the id of already upserted vector, use `Index.query_by_id()`
    ///
    /// Args:
    ///     top_k (int): The number of results to return for each query.
    ///     values (Optional[List[float]]): The values for a new, unseen query vector. This should be the same length as the dimension of the index being queried. The results will be the `top_k` vectors closest to the given vector. Can not be used together with `id`.
    ///     sparse_values (Optional[SparseValues]): The query vector's sparse values.
    ///     namespace (Optional[str]): Optional namespace in which vectors will be queried.
    ///     filter (Optional[dict]): The filter to apply. You can use vector metadata to limit your search. See <https://www.pinecone.io/docs/metadata-filtering/>
    ///     include_values (bool): Indicates whether vector values are included in the response.
    ///     include_metadata (bool): Indicates whether metadata is included in the response as well as the ids.
    ///
    /// Returns:
    ///     list of QueryResults
    #[allow(clippy::too_many_arguments)]
    pub fn query(
        &mut self,
        top_k: i32,
        values: Option<Vec<f32>>,
        sparse_values: Option<core_data_types::SparseValues>,
        namespace: &str,
        filter: Option<BTreeMap<String, core_data_types::MetadataValue>>,
        include_values: bool,
        include_metadata: bool,
    ) -> PineconeResult<Vec<core_data_types::QueryResult>> {
        if top_k < 1 {
            return Err(core_error::ValueError("top_k must be greater than 0".to_string()).into());
        }
        let res = self.runtime.block_on(self.inner.query(
            namespace,
            values,
            sparse_values,
            top_k as u32,
            filter,
            include_values,
            include_metadata,
        ))?;
        Ok(res)
    }

    #[pyo3(signature = (id, top_k, namespace="", filter=None, include_values=false, include_metadata=false))]
    #[pyo3(
        text_signature = "($self, id, top_k, namespace='', filter=None, include_values=False, include_metadata=False)"
    )]
    /// Query by id
    ///
    /// The `Query by id` operation searches a namespace given the `id` of a vector already residing in the Index.
    /// It retrieves the ids of the most similar items in a namespace, along with their similarity scores.
    /// To query by new unseen vector use `Index.query()`
    ///
    /// Args:
    ///     id (str): An id of a vector already upserted to the relevant namespace. The results will be the `top_k` nearest neighbours of the vector with the given id. Cannot be used together with `values`.
    ///     top_k (int): The number of results to return for each query.
    ///     namespace (Optional[str]): Optional namespace in which vectors will be queried.
    ///     filter (Optional[dict]): The filter to apply. You can use vector metadata to limit your search. See <https://www.pinecone.io/docs/metadata-filtering/>
    ///     include_values (bool): Indicates whether vector values are included in the response.
    ///     include_metadata (bool): Indicates whether metadata is included in the response as well as the ids.
    ///
    /// Returns:
    ///     list of QueryResults
    pub fn query_by_id(
        &mut self,
        id: &str,
        top_k: i32,
        namespace: &str,
        filter: Option<BTreeMap<String, core_data_types::MetadataValue>>,
        include_values: bool,
        include_metadata: bool,
    ) -> PineconeResult<Vec<core_data_types::QueryResult>> {
        if top_k < 1 {
            return Err(core_error::ValueError("top_k must be greater than 0".to_string()).into());
        }
        let res = self.runtime.block_on(self.inner.query_by_id(
            namespace,
            id,
            top_k as u32,
            filter,
            include_values,
            include_metadata,
        ))?;
        Ok(res)
    }

    #[pyo3(signature = (filter=None))]
    #[pyo3(text_signature = "(filter=None)")]
    /// Describe index stats.
    ///
    /// The `DescribeIndexStats` operation returns the number of vectors present in the index, for all the namespaces
    /// and the fullness of the index. Can also accept a filter to count the number of vectors matching the filter.
    ///
    /// Args:
    ///     filter (Dict[str, Union[str, float, int, bool, List, dict]]):
    ///     If this parameter is present, the operation only returns statistics for vectors that satisfy the filter.
    ///     See https://www.pinecone.io/docs/metadata-filtering/.. [optional]
    ///
    /// Returns:
    ///     An `IndexStats` object containing index statistics.
    pub fn describe_index_stats(
        &mut self,
        filter: Option<BTreeMap<String, core_data_types::MetadataValue>>,
    ) -> PineconeResult<core_data_types::IndexStats> {
        let res = self
            .runtime
            .block_on(self.inner.describe_index_stats(filter))?;
        Ok(res)
    }

    #[pyo3(signature = (ids, namespace=""))]
    #[pyo3(text_signature = "($self, ids, namespace='')")]
    /// Fetch
    ///
    /// The fetch operation looks up and returns vectors, by ID, from a single namespace.
    /// The returned vectors include the vector data and/or metadata.
    ///
    /// Args:
    ///     ids (List[str]): The vector IDs to fetch.
    ///     namespace (str): The namespace to fetch vectors from.
    ///                      If not specified, the default namespace is used. [optional]
    ///
    /// Examples:
    ///     >>> index.fetch(ids=['id1', 'id2'], namespace='my_namespace')
    ///     >>> index.fetch(ids=['id1', 'id2'])
    ///
    /// Returns: a dictionary of vector IDs to the fetched vectors.
    pub fn fetch(
        &mut self,
        ids: Vec<String>,
        namespace: &str,
    ) -> PineconeResult<HashMap<String, core_data_types::Vector>> {
        let res = self.runtime.block_on(self.inner.fetch(namespace, &ids))?;
        Ok(res)
    }

    #[pyo3(signature = (id, values=None, sparse_values=None, set_metadata=None, namespace=""))]
    #[pyo3(
        text_signature = "($self, id, values=None, sparse_values=None, set_metadata=None, namespace='')"
    )]
    /// Update
    /// The Update operation updates vector in a namespace.
    /// If a value is included, it will overwrite the previous value.
    /// If a set_metadata is included,
    /// the values of the fields specified in it will be added or overwrite the previous value.
    ///
    /// Examples:
    ///     >>> index.update(id='id1', values=[1, 2, 3], namespace='my_namespace')
    ///     >>> index.update(id='id1', set_metadata={'key': 'value'}, namespace='my_namespace')
    ///     >>> index.update(id='id1', values=[1, 2, 3], sparse_values=SparseValues(indices=[1, 2], values=[0.2, 0.4]),
    ///                      namespace='my_namespace')
    ///
    /// Args:
    ///     id (str): Vector's unique id.
    ///     values (List[float]): vector values to set. [optional]
    ///     sparse_values: (SparseValues): sparse values to update for the vector.
    ///     set_metadata (Dict[str, Union[str, float, int, bool, List[str]]]]): metadata to set for vector. [optional]
    ///     namespace (str): Namespace name where to update the vector.. [optional]
    ///
    pub fn update(
        &mut self,
        id: &str,
        values: Option<Vec<f32>>,
        sparse_values: Option<core_data_types::SparseValues>,
        set_metadata: Option<BTreeMap<String, core_data_types::MetadataValue>>,
        namespace: &str,
    ) -> PineconeResult<()> {
        self.runtime.block_on(self.inner.update(
            id,
            values.as_ref(),
            sparse_values,
            set_metadata,
            namespace,
        ))?;
        Ok(())
    }

    #[pyo3(signature = (ids, namespace=""))]
    #[pyo3(text_signature = "($self, ids, namespace='')")]
    /// Delete
    /// Delete vectors by ID from a given namespace.
    ///
    /// Args:
    ///     ids (List[str]): A list of IDs for vectors to be deleted.
    ///     namespace (str): The name of the namespace from which vectors will be deleted. If None, the default namespace will be used.
    ///
    /// Returns:
    ///    None
    pub fn delete(&mut self, ids: Vec<String>, namespace: &str) -> PineconeResult<()> {
        self.runtime.block_on(self.inner.delete(ids, namespace))?;
        Ok(())
    }

    #[pyo3(signature = (filter, namespace=""))]
    #[pyo3(text_signature = "($self, filter, namespace='')")]
    /// Delete by filter
    /// The delete by filter operation deletes a list of vectors from a given namespace that match the filter.
    ///
    /// Args:
    ///     filter (Dict[str, Union[str, float, int, bool, List, dict]]): filter to be applied to delete the vectors. See https://www.pinecone.io/docs/metadata-filtering/
    ///     namespace (Optional[str]): The name of the namespace from which vectors will be deleted. If None, the default namespace will be used.
    ///
    /// Returns:
    ///    None
    pub fn delete_by_metadata(
        &mut self,
        filter: Option<BTreeMap<String, core_data_types::MetadataValue>>,
        namespace: &str,
    ) -> PineconeResult<()> {
        self.runtime
            .block_on(self.inner.delete_by_metadata(filter, namespace))?;
        Ok(())
    }

    #[pyo3(signature = (namespace=""))]
    #[pyo3(text_signature = "($self, namespace='')")]
    /// Delete all
    /// The delete all operation deletes all the vectors from a given namespace.
    ///
    /// Args:
    ///     namespace (str): The name of the namespace from which vectors will be deleted. If None, the default namespace will be used.
    ///
    /// Returns:
    ///    None
    pub fn delete_all(&mut self, namespace: &str) -> PineconeResult<()> {
        self.runtime.block_on(self.inner.delete_all(namespace))?;
        Ok(())
    }
}
